### ISSUE TITLE
handling of attributes and SP interoperability

### ISSUE THREAD

**Issue by dub357** [simplesamlphp/simplesamlphp#690]
SimpleSamlPhp has a few major drawbacks when it comes to attribute handling and acting as an IdP proxy (authsource = saml:SP).  Basically this boils down to SimpleSamlPhp "losing" attribute information from the original assertion....the friendly name and the name format.
First issue is with name formats.  The upstream IdP may have passed back attributes with different name formats, but SimpleSamlPhp uses the exact same name format in the assertion passed by to the SP.  Whats worse, if there is no name format specified in any of its metadata, the 'basic' format is used for every attribute.  The Shibboleth SP implementation by default assumes all attributes will be of NameFormat URI (urn:oasis:names:tc:SAML:2.0:attrname-format:uri) unless they are explicitly mapped out to be something different.  So unless the Shib SP config explicitly declares a name format for every attribute is expects, no attributes from SimpleSamlPhp are processed.  Even if it does specify name formats for every attribute, many of them could be incorrect because SimpleSamlPhp sets the same name format for every attribute it releases.  
The second issue is with attribute friendly names.  SimpleSamlPhp simply does NOT send them.  Instead it has this concept of authproc filters which trying to work around this issue... converting what would be a friendly name to a 'federated' name or vice-versa.  Most IdPs send attributes with names that aren't "human readable" (like OIDs or others that are basically a URI/URN) and authproc filters convert these to a "friendly name".   However for SP implementations that rely on both a name and a friendly name, this can be problematic.  

So to sum up the feature request, when acting as a proxy (authsource saml:SP), SimpleSamlPhp should internally remember the name format and friendly name of the attributes received from the upstream IdP so it can pass them back to the original SP.

**Comment by jaimeperez** [comment]
Hi @dub357!

Well, the thing is... proxy mode is not officially supported (and never has). I get it's confusing because due to the way SimpleSAMLphp is written, it just works as a proxy without much trouble. However, proxy mode has never been officially supported and that's why you may observe these issues as well as others (i.e. lack of downstream SLO propagation).

In this particular case, I think the issues you are facing should be resolved by configuration. In the end, it is your proxy (and not the upstream IdP) who is connected to the SP (and the other way around), so your proxy should know what name format is needed by that particular SP, and the configuration should be done accordingly.

Regarding the `FriendlyName` of attributes, I have never seen any single SP that uses it for anything other than displaying a _friendly name_ for the attribute to the users, and I actually believe processing them in any other way would be against the spec.

In any case, even though proxy mode is not really supported, we are glad to make SimpleSAMLphp better, and therefore contributions are always welcome. Would you be willing to submit a couple of pull requests to fix the issues you are mentioning?

**Comment by dub357** [comment]
Strange that you say the proxy mode (authsource=saml:SP) isn't officially supported when its clearly in the documentation and not once stated that its "not supported/experimental, etc":
https://simplesamlphp.org/docs/stable/saml:sp

According to the spec, different attributes can have different name formats.  SimpleSamlPhp doesn't differentiate...all attributes "share" the same name format.  An IdP could be passing back "urn:oid:1.3.6.1.4.1.5923.1.1.1.6" (eduPersonPrincipalName) which is an attribute with a URI name format and a customized 'basic' attribute, but when SimpleSamlPhp responds back to the SP, both attributes will get the exact same name format.  Whats worse, if the name format isn't explicitly set (not known), it defaults to "urn:oasis:names:tc:SAML:2.0:attrname-format:basic" when it probably would make more sense to use "urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified".  

As for the FriendlyName, you are correct that having that is merely a way to display a human-readable name for an attribute, but it also allows an SP to not have to know up-front the "friendly" name for every single attribute that it will ever get passed.  Most SPs need some way to convert an attribute received in an assertion to a session-based attribute that can be used for principal naming, authorization checks, etc.  The Shibboleth SP has an attribute mapping function to facilitate this, but again - it requires predefined configuration and if an IdP releases an attribute that isn't in its attribute map config, it ignores it.  Many SP implementations don't have this functionality to map attributes to some other name and simply store all attributes from the assertion in the session.  They do have the ability to prefer the friendly name of attributes over the often non-understandable 'real' name and thats when the friendly name becomes very useful.

I wouldn't mind submitting pull requests...I do have another issue outstanding that is an easy fix and I also have what would be an enhancement that a number of other folks of asked for as well...

**Comment by dub357** [comment]
Looks like I may have put this issue in the wrong place....should it go here?
https://github.com/simplesamlphp/saml2/issues

**Comment by jaimeperez** [comment]
The documentation you are looking at has nothing to do with proxy mode. That's the documentation for the SAML SP authentication source, what applications use to deploy a Service Provider. Incidentally, if you want to run a proxy, you would configure SimpleSAMLphp as an IdP whose authentication source is precisely a SAML SP, but that doesn't mean that the SP authn source is intended for its use in proxies. Note that the word "proxy" is not mentioned any single time in the document.

Regarding the common `NameFormat` for all attributes, you are absolutely right. Currently, there's no way to send attributes out with different formats. Don't know if that's a good idea or something any SP accepts, though, but you are right anyway.

Not sure whether this issue is right here or should be in simplesamlphp/saml2. It depends a bit on how you plan to solve it, and I don't have all the code in my head right now (and can't really look into it as I'm in a workshop for the whole week), but I'd say this will need a lot of work through the internals of SimpleSAMLphp, so it's probably fine here. The thing is... when you get a `SAMLResponse`, the assertion with the attributes is parsed, and its contents are used to bootstrap a session. Internally, SSP doesn't keep any information about attribute NameFormats, and it won't be possible to do that without having the SP and the IdP part communicate to each other (which they don't do now, precisely the reason why proxy mode is not officially supported) so you'll probably need to look at the attribute names and _guess_ the formats from there. Also, keep in mind that this is the IdP functionality you would be changing (not anything proxy-specific), and therefore it needs to be backwards-compatible and transparent for those **not** using it as a proxy.

**Comment by dub357** [comment]
Ok - I have a "fix" for this in my fork.  Still need to run tests against it, to make sure its coverage is complete but seems to work when I test in the browser using multiple auth sources.  Required changes in both the "saml2" project to store attributes as SAML2\XML\saml\Attribute objects instead of arrays of strings/ints/DOMElements (with backwards compatibility for getting attributes the old way by default), and a change to the "simplesamlphp" project for dealing with attributes from upstream IdPs in the new fashion and ability to specify config arrays which map attribute names to nameFormats and friendlyNames.  
Once I verify that the new code passes all tests, I'll generate a pull request

**Comment by jaimeperez** [comment]
That sounds great @dub357!

I think that would make a nice new feature in 2.0, because then we can have richer semantics for every attribute, apart from solving this particular issue.

Really looking forward to your PR 😄

**Comment by dub357** [comment]
So I'm trying to get back into this after life getting in the way and I'm trying to run the tests for simplesamlphp (saml2 is passing).  However, when running composer to install the dependencies, I'm getting errors due to PHP versions.  I'm running 5.4.16 (setup by my system admins) and I'm seeing at least 10 problems related to now requiring version 5.5 or so.  The 1.15 upgrade docs in master state the minimum requirement is now 5.4...is this not the case?

**Comment by jaimeperez** [comment]
Hi @dub357!

PHP 5.4 is the current minimum requirement, not 5.5. How are you pulling the dependencies? What are the errors you are getting?

**Comment by dub357** [comment]
All I did was fork the repo, make my changes (no dependencies, only 8 files:).
I then run the following command from within the simplesamlphp directory:

`php ../composer.phar install --dry-run`

And I get the following output.  Looks like the 5.5 requirement is based on other dependencies:

```Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for phpdocumentor/reflection-common 1.0 -> satisfiable by phpdocumentor/reflection-common[1.0].
    - phpdocumentor/reflection-common 1.0 requires php >=5.5 -> your PHP version (5.4.16) does not satisfy that requirement.
  Problem 2
    - Installation request for phpdocumentor/reflection-docblock 3.2.0 -> satisfiable by phpdocumentor/reflection-docblock[3.2.0].
    - phpdocumentor/reflection-docblock 3.2.0 requires php >=5.5 -> your PHP version (5.4.16) does not satisfy that requirement.
  Problem 3
    - Installation request for phpdocumentor/type-resolver 0.4.0 -> satisfiable by phpdocumentor/type-resolver[0.4.0].
    - phpdocumentor/type-resolver 0.4.0 requires php ^5.5 || ^7.0 -> your PHP version (5.4.16) does not satisfy that requirement.
  Problem 4
    - Installation request for symfony/config v3.3.6 -> satisfiable by symfony/config[v3.3.6].
    - symfony/config v3.3.6 requires php >=5.5.9 -> your PHP version (5.4.16) does not satisfy that requirement.
  Problem 5
    - Installation request for symfony/console v3.3.6 -> satisfiable by symfony/console[v3.3.6].
    - symfony/console v3.3.6 requires php >=5.5.9 -> your PHP version (5.4.16) does not satisfy that requirement.
  Problem 6
    - Installation request for symfony/debug v3.3.6 -> satisfiable by symfony/debug[v3.3.6].
    - symfony/debug v3.3.6 requires php >=5.5.9 -> your PHP version (5.4.16) does not satisfy that requirement.
  Problem 7
    - Installation request for symfony/filesystem v3.3.6 -> satisfiable by symfony/filesystem[v3.3.6].
    - symfony/filesystem v3.3.6 requires php >=5.5.9 -> your PHP version (5.4.16) does not satisfy that requirement.
  Problem 8
    - Installation request for symfony/stopwatch v3.3.6 -> satisfiable by symfony/stopwatch[v3.3.6].
    - symfony/stopwatch v3.3.6 requires php >=5.5.9 -> your PHP version (5.4.16) does not satisfy that requirement.
  Problem 9
    - Installation request for symfony/yaml v3.3.6 -> satisfiable by symfony/yaml[v3.3.6].
    - symfony/yaml v3.3.6 requires php >=5.5.9 -> your PHP version (5.4.16) does not satisfy that requirement.
  Problem 10
    - phpdocumentor/reflection-docblock 3.2.0 requires php >=5.5 -> your PHP version (5.4.16) does not satisfy that requirement.
    - phpspec/prophecy v1.7.0 requires phpdocumentor/reflection-docblock ^2.0|^3.0.2 -> satisfiable by phpdocumentor/reflection-docblock[3.2.0].
    - Installation request for phpspec/prophecy v1.7.0 -> satisfiable by phpspec/prophecy[v1.7.0].
```
`

**Comment by tvdijen** [comment]
I've been trying to reproduce this, but I just can't...

Could you try clearing your composer-cache?
`rm -r ~/.composer`

**Comment by jaimeperez** [comment]
I can't reproduce it either as I'm running PHP 5.6. However, I see the following:

- phpdocumentor/reflection-docblock 2.0.5 (last 2.x version) requires PHP >= 5.3.3.
- phpdocumentor/reflection-dockblock 3.2.0 requires PHP >= 5.5.

The build is working with PHP 5.4 in travis, which leads me to think this is a local issue. It looks also like there's something wrong with composer because it's trying to install the unsatisfiable 3.0.2 version and not the 2.0.5 version, which is satisfiable, and also satisfies the requirement by phpspec/prophecy. Apart from clearing the composer cache, can you make sure you have an updated version of composer?

**Comment by tvdijen** [comment]
@jaimeperez : I'm not running 5.4 either, but your trick composer to think you're running a certain PHP-version:
`    "config": {
        "platform": {
            "php": "5.4"
        }
    },
`

**Comment by dub357** [comment]
I'm new to composer so this is a fresh install with the latest version: 1.5.2.  So far I have only used it to load dependencies for the saml2 project (which worked fine).  Its only the simplesamlphp project that has this issue.
Anyway - clearing the composer cache does not fix the problem nor does adding the following to the simplesamlphp project composer.json config file:

`"config": { "platform": { "php": "5.6" } }`

The only thing that allows me to get around this problem is removing the composer.lock that was committed to the project.  This is because that file is what defines the minimum php version for dependencies.  However, once I do this and run composer, I then get a new composer.lock file.  I guess I can simply just not commit that.  At least I have the dependencies installed now and can continue

**Comment by dub357** [comment]
So after finally getting past the composer dependency issue, I've run into the same problem that I ran into with the saml2 project.  Basically is comes down to the unit tests that compare strings for equality.  A ton of tests use the "assertEquals" methods to check if a string is equal to another string.  These fail frequently for me because of PHP_EOL.  I'm running on Linux which defines the PHP_EOL character to be "\n" but the test code looks to be written on a Windows box because it contains "\r\n" in the strings that it tries to compare with.  Well - the generated strings on my Linux box don't contain the "\r" character so the test fails.  
Seems that this issue has been "corrected" in newer versions of phpunit, but I think we are stuck on 4.8 because it works with PHP 5.3-5.6.  The newer versions require a newer version of PHP - which exceeds the minimum PHP version of simplesamlphp.  
I simply ignored these issues when running the tests for saml2 and it seems like I'll be doing the same for simplesamlphp as well.  Has this issue not come up with other developers before?

**Comment by dub357** [comment]
Hmm - this may be my issue...I used a windows box (which had a Git client) on it to clone the forked repo and then began my changes/testing on a Linus box.  I think during the cloning that the \r\n line endings were introduced.

**Comment by pradtke** [comment]
> Regarding the FriendlyName of attributes, I have never seen any single SP that uses it for anything other than displaying a friendly name for the attribute to the users, and I actually believe processing them in any other way would be against the spec.

FWIW, this week we've encountered two SPs, both in InCommon, that require FriendlyName to be asserted.

**Comment by thijskh** [comment]
Just as a side note, the minimum PHP requirement is now 5.5 so at least some of the issues mentioned above should no longer be relevant. Still needs someone to work on the saml2 PR though.
